### PR TITLE
feat: flashloans

### PR DESCRIPTION
**Pull request by zavgorodnii**
This PR adds the `neutron-flashloans` contract that uses another address as the source of funds (in our case, the DAO core address). You can find a detailed description of the contract logic and implementation in README.md, or in the docs.

Note: the `neutron-flashloans-user` contract that is only used in the integration tests is added to the `neutron-dev-contracts` repo: https://github.com/neutron-org/neutron-dev-contracts/pull/57.

- [x] **Integration tests:** https://github.com/neutron-org/neutron-integration-tests/pull/304/files
- [x] **Integration tests run:** https://github.com/neutron-org/neutron-tests/actions/runs/12966875346
- [x] **Documentation:** https://github.com/neutron-org/neutron-docs/pull/177

### CODE REVIEW

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+/// Returns the list of current balances on the contract's account for the coins that were
+/// requested by the borrower. If any actual coin balance is lower than the requested amount,
+/// returns an error.
+fn get_pre_loan_balances(
+    deps: &DepsMut,
+    source: Addr,
+    requested_amount: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
```
Comment (pr0n00gler) [inline contracts/dao/neutron-flashloans/src/contract.rs:197]: I don't think this is safe.

I can spam the `source` address with an _infinite_ amount of TF tokens, so after the certain number of created tokens your `AllBalances` query will return a gas error, thus users can't get new flashloans

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+        address: source.to_string(),
+    };
+    // Get the response (all balances)
+    let all_balances_response: AllBalanceResponse =
+        deps.querier.query(&all_balances_query.into())?;
+
+    // Filter all balances leaving only the balances of the requested coins
+    let mut pre_loan_balances: Vec<Coin> = vec![];
+    for requested_coin in requested_amount {
```
Comment (pr0n00gler) [inline contracts/dao/neutron-flashloans/src/contract.rs:206]: Or i can get a gas error here.

TLDR infinite loops are not good

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+}
+
+// Sums the pre_loan_balances with the fee.
+// WARNING: this function assumes that the input vectors are of the same length,
+// and that the order of the denoms is the same.
+fn calculate_expected_balances(
+    pre_loan_balances: Vec<Coin>,
+    fee: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    let mut expected_balances: Vec<Coin> = vec![];
```
Comment (pr0n00gler) [inline contracts/dao/neutron-flashloans/src/contract.rs:249]:
```suggestion
    let mut expected_balances: Vec<Coin> = Vec::with_capacity(pre_loan_balances.len());
```

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+    Ok(pre_loan_balances)
+}
+
+/// Calculates the fee by multiplying each of the requested assets by fee_rate.
+fn calculate_fee(
+    requested_amount: Vec<Coin>,
+    fee_rate: Decimal,
+) -> Result<Vec<Coin>, ContractError> {
+    let mut fee: Vec<Coin> = vec![];
```
Comment (pr0n00gler) [inline contracts/dao/neutron-flashloans/src/contract.rs:233]:
```suggestion
    let mut fee: Vec<Coin> = Vec::with_capacity(requested_amount.len());
```

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+// Loads the information about the current loan, returns an error if the information
+// is missing.
+fn must_get_active_loan(deps: &DepsMut) -> Result<ActiveLoan, ContractError> {
+    let active_loan = ACTIVE_LOAN.load(deps.storage)?;
+    if active_loan.is_none() {
+        return Err(ContractError::UnexpectedState {});
+    }
+
+    Ok(active_loan.unwrap().clone())
```
Comment (pr0n00gler) [inline contracts/dao/neutron-flashloans/src/contract.rs:384]: Unnecessary `.clone()` call

```suggestion
    Ok(active_loan.unwrap())
```

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+    }
+}
+
+fn check_expected_balances(
+    deps: Deps,
+    source: Addr,
+    expected_balances: Vec<Coin>,
+) -> Result<(), ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
```
Comment (pr0n00gler) [inline contracts/dao/neutron-flashloans/src/contract.rs:352]: The same about a potentially _infinite_ loop

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: contracts/dao/neutron-flashloans/src/contract.rs (hunk 2 above)
Comment (zavgorodnii) [reply contracts/dao/neutron-flashloans/src/contract.rs:206]: I see your point, spamming might be an issue; will fix.

File: contracts/dao/neutron-flashloans/src/contract.rs (hunk 1 above)
Comment (zavgorodnii) [reply contracts/dao/neutron-flashloans/src/contract.rs:197]: Do you suggest having an allow list of tokens that we are ready to loan?

File: contracts/dao/neutron-flashloans/src/contract.rs (hunk 6 above)
Comment (zavgorodnii) [reply contracts/dao/neutron-flashloans/src/contract.rs:352]: Acknowledged, will fix!

File: contracts/dao/neutron-flashloans/src/contract.rs (hunk 1 above)
Comment (pr0n00gler) [reply contracts/dao/neutron-flashloans/src/contract.rs:197]: Either that, or do a separate `BankQuery::Balance { address: String, denom: String }` query.
That way query can be broken only if a user himself provides a lot of tokens to loan. But it's kinda his problem, not contract's

Comment (zavgorodnii) [reply contracts/dao/neutron-flashloans/src/contract.rs:197]: Makes sense, I like the separate query approach more

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+
+        // If the source doesn't have (enough of) the requested coin, return an error
+        if maybe_source_coin.is_none() {
+            return Err(ContractError::InsufficientFunds {
+                denom: requested_coin.denom,
+            });
+        }
+
+        let source_coin = maybe_source_coin.unwrap();
+        pre_loan_balances.push(source_coin.clone())
```
Comment (sotnikov-s) [inline contracts/dao/neutron-flashloans/src/contract.rs:222]: this can be rewritten in a cleaner way and without upwrap usage
```suggestion
        match maybe_source_coin {
            Some(coin) => pre_loan_balances.push(coin.clone()),
            // If the source doesn't have (enough of) the requested coin, return an error
            None => {
                return Err(ContractError::InsufficientFunds {
                    denom: requested_coin.denom,
                })
            }
        }
```

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+        ))
+    }
+
+    Ok(expected_balances)
+}
+
+/// A simple function to build the (stargate -> authz -> bank) /cosmos.bank.v1beta1.MsgSend message.
+fn get_stargate_authz_bank_send_msg(
+    env: Env,
+    config: Config,
```
Comment (sotnikov-s) [inline contracts/dao/neutron-flashloans/src/contract.rs:263]: would be cleaner here to pass only the source, not the whole config

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+    amount: Vec<Coin>,
+) -> CosmosMsg<NeutronMsg> {
+    // First we create the bank MsgSend
+    let bank_send_msg = MsgSend {
+        from_address: config.source.to_string(),
+        to_address: borrower.to_string(),
+        amount: amount
+            .iter()
+            .map(|x| ProtoCoin {
+                denom: x.clone().denom,
```
Comment (sotnikov-s) [inline contracts/dao/neutron-flashloans/src/contract.rs:274]: why clone the whole coin instance?
```suggestion
                denom: x.denom.clone(),
```

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+            // If we are here, the money is already on the borrower's account. This means that
+            // we can proceed to call the borrower's handler.
+
+            let config: Config = CONFIG.load(deps.storage)?;
+            let active_loan = must_get_active_loan(&deps)?;
+
+            let msg = CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: active_loan.borrower.to_string(),
+                msg: to_json_binary(&BorrowerInterface::ProcessLoan {
+                    // We sent the return address to the source address
```
Comment (sotnikov-s) [inline contracts/dao/neutron-flashloans/src/contract.rs:312]:
```suggestion
                    // We set the return address to the source address
```
btw is this comment even necessary? it just literally explains the field assignment operation saying nothing about the motivation (which must be obvious)

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+                    return_address: config.source,
+                    loan_amount: active_loan.amount,
+                    fee: active_loan.fee,
+                })
+                .unwrap(),
```
Comment (sotnikov-s) [inline contracts/dao/neutron-flashloans/src/contract.rs:317]:
```suggestion
                })?,
```

File: contracts/dao/neutron-flashloans/src/contract.rs [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+        }),
+    )?;
+
+    // Send a (stargate -> authz -> bank) /cosmos.bank.v1beta1.MsgSend submessage with
+    // reply_on_success strategy (we want the transaction to be simply reverted in case of an
+    // error).
+    let msg_send = get_stargate_authz_bank_send_msg(env, config, info.sender, amount);
+    Ok(Response::new()
+        .add_submessage(SubMsg::reply_on_success(msg_send, AUTHZ_BANK_SEND_REPLY_ID))
+        .add_attribute("action", "execute_get_loan"))
```
Comment (sotnikov-s) [inline contracts/dao/neutron-flashloans/src/contract.rs:167]:
```suggestion
        .add_attribute("action", "execute_request_loan"))
```

**Review by sotnikov-s — CHANGES_REQUESTED** [review]
please don't forget to move the flashloan user contract to the dev contracts repo

**Review by pr0n00gler — APPROVED** [review]

**Review by sotnikov-s — CHANGES_REQUESTED** [review]
all but this tiny issue LGTM

**Review by sotnikov-s — APPROVED** [review]
